chapter_natural-language-processing-pretraining/word-embedding-dataset.ipynb

```diff
@@ -0,0 +1,825 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static Pair<ArrayList<Integer>, ArrayList<ArrayList<Integer>>> getCentersAndContext(\n",
+    "        Integer[][] corpus, int maxWindowSize) {\n",
+    "    ArrayList<Integer> centers = new ArrayList<>();\n",
+    "    ArrayList<ArrayList<Integer>> contexts = new ArrayList<>();\n",
+    "\n",
+    "    for (Integer[] line : corpus) {\n",
+    "        // Each sentence needs at least 2 words to form a \"central target word\n",
+    "        // - context word\" pair\n",
+    "        if (line.length < 2) {\n",
+    "            continue;\n",
+    "        }\n",
+    "        centers.addAll(Arrays.asList(line));\n",
+    "        for (int i = 0; i < line.length; i++) { // Context window centered at i\n",
+    "            int windowSize = new Random().nextInt(maxWindowSize - 1) + 1;\n",
+    "            List<Integer> indices =\n",
+    "                    IntStream.range(\n",
+    "                                    Math.max(0, i - windowSize),\n",
+    "                                    Math.min(line.length, i + 1 + windowSize))\n",
+    "                            .boxed()\n",
+    "                            .collect(Collectors.toList());\n",
+    "            // Exclude the central target word from the context words\n",
+    "            indices.remove(indices.indexOf(i));\n",
+    "            ArrayList<Integer> context = new ArrayList<>();\n",
+    "            for (Integer idx : indices) {\n",
+    "                context.add(line[idx]);\n",
+    "            }\n",
+    "            contexts.add(context);\n",
+    "        }\n",
+    "    }\n",
+    "    return new Pair<>(centers, contexts);\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 19
+   },
+   "source": [
+    "Next, we create an artificial dataset containing two sentences of 7 and 3 words, respectively. Assume the maximum context window is 2 and print all the central target words and their context words.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "Integer[][] tinyDataset =\n",
+    "        new Integer[][] {\n",
+    "            IntStream.range(0, 7)\n",
+    "                    .boxed()\n",
+    "                    .collect(Collectors.toList())\n",
+    "                    .toArray(new Integer[] {}),\n",
+    "            IntStream.range(7, 10)\n",
+    "                    .boxed()\n",
+    "                    .collect(Collectors.toList())\n",
+    "                    .toArray(new Integer[] {})\n",
+    "        };\n",
+    "\n",
+    "System.out.println(\"dataset \" + Arrays.deepToString(tinyDataset));\n",
+    "Pair<ArrayList<Integer>, ArrayList<ArrayList<Integer>>> centerContextPair =\n",
+    "        getCentersAndContext(tinyDataset, 2);\n",
+    "for (int i = 0; i < centerContextPair.getValue().size(); i++) {\n",
```

Comment:
the result here doesn't match d2l book

Reply:
getCentersAndContext/get_centers_and_contexts contains a random variable which affects results. When creating the section I did the value of windowSize constant to test and results were indeed the same.